src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs

```diff
@@ -340,14 +407,49 @@ private void Write(RuntimeModel.IObjectWriter writer, bool hashing, Action<Packa
             foreach (var pair in _projects)
             {
                 var project = pair.Value;
+                WriteProject(writer, hashing, writeAction, project, projectNameToHashCode);
+            }
+
+            writer.WriteObjectEnd();
+            writer.WriteObjectEnd();
+        }
+
+        private void WriteProject(IObjectWriter writer, bool hashing, Action<PackageSpec, IObjectWriter, bool, IEnvironmentVariableReader> writeAction, PackageSpec project, Dictionary<string, string>? projectNameToHashCode)
+        {
+            if (hashing && projectNameToHashCode != null)
+            {
+                string? projectHash = null;
+
+                lock (projectNameToHashCode)
+                {
+                    projectNameToHashCode.TryGetValue(project.RestoreMetadata.ProjectUniqueName, out projectHash);
+                }
+
+                if (projectHash == null)
+                {
+                    using IHashFunction hashFunc = UseLegacyHashFunction ? new Sha512HashFunction() : new FnvHash64Function();
+                    using var projectWriter = new HashObjectWriter(hashFunc);
+                    writeAction.Invoke(project, projectWriter, hashing, EnvironmentVariableWrapper.Instance);
+                    projectHash = projectWriter.GetHash();
+                }
+
+                lock (projectNameToHashCode)
+                {
+                    if (!projectNameToHashCode.ContainsKey(project.RestoreMetadata.ProjectUniqueName))
```

Comment:
can this lock be moved inside the `if (projectHash == null)` block, so it doesn't need to run on cache hits?

Reply:
Yeah, the 2nd check would help.